src/core/server/metrics/integration_tests/server_collector.test.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Subject } from 'rxjs';
+import { take } from 'rxjs/operators';
+import supertest from 'supertest';
+import { Server as HapiServer } from 'hapi';
+import { createHttpServer } from '../../http/test_utils';
+import { HttpService, IRouter } from '../../http';
+import { contextServiceMock } from '../../context/context_service.mock';
+import { ServerMetricsCollector } from '../collectors/server';
+
+describe('ServerMetricsCollector', () => {
+  let server: HttpService;
+  let collector: ServerMetricsCollector;
+  let hapiServer: HapiServer;
+  let router: IRouter;
+
+  const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
+  const sendGet = (path: string) => supertest(hapiServer.listener).get(path);
+
+  beforeEach(async () => {
+    server = createHttpServer();
+    const contextSetup = contextServiceMock.createSetupContract();
+    const httpSetup = await server.setup({ context: contextSetup });
+    hapiServer = httpSetup.server;
+    router = httpSetup.createRouter('/');
+    collector = new ServerMetricsCollector(hapiServer);
+  });
+
+  afterEach(async () => {
+    await server.stop();
+  });
+
+  it('collect requests infos', async () => {
+    router.get({ path: '/', validate: false }, async (ctx, req, res) => {
+      return res.ok({ body: '' });
+    });
+    await server.start();
+
+    let metrics = await collector.collect();
+
+    expect(metrics.requests).toEqual({
+      total: 0,
+      disconnects: 0,
+      statusCodes: {},
+    });
+
+    await sendGet('/');
+    await sendGet('/');
+    await sendGet('/not-found');
+
+    metrics = await collector.collect();
+
+    expect(metrics.requests).toEqual({
+      total: 3,
+      disconnects: 0,
```

Comment:
I believe disconnects is untested

Reply:
Yea it is, it seems there is no way to force a disconnect on a `supertest` connection...

Reply:
Erata: There is, added test.